American Option Dynamic Prog.cpp

```diff
@@ -0,0 +1,148 @@
+// Calculating the price of an American Option using Dynamic Programming
+//where the Probability Matrix is derived from the Trinomial Lattice.
+
+#include <iostream>
+#include <iomanip>
+#include <cmath>
+#include <fstream>
+#include <cstdlib>
+#include <algorithm>
+#include ".../newmat10/newmat.h"
+using namespace std;
+
+float up_factor, uptick_prob, downtick_prob, notick_prob, risk_free_rate, strike_price;
+float initial_stock_price, expiration_time, volatility, R;
+int no_of_divisions;
+
+float max(float a, float b) {
+	return (b < a )? a:b;
+}
+
+float american_call_option_dyn_prog()
+{
+
+	// create the probability matrix
+	Matrix transition_probability(2*no_of_divisions+1, 2*no_of_divisions+1);
+	
+	for (int i = 1; i <= 2*no_of_divisions+1; i++)
+		for (int j = 1; j <= 2*no_of_divisions+1; j++)
+			transition_probability(i,j) = 0.0;
+	
+	// boundary values of the probabilities need to be entered
+	transition_probability(1,1) = 1.0 - uptick_prob;
+	transition_probability(1,2) = uptick_prob;
+	transition_probability(2*no_of_divisions+1,2*no_of_divisions+1) = 1.0 - downtick_prob;
+	transition_probability(2*no_of_divisions+1,2*no_of_divisions) = downtick_prob;
+	
+	for (int i = 2; i <= 2*no_of_divisions; i++)
+		for (int j = 1; j <= 2*no_of_divisions+1; j++) {
+			if (j == (i-1))
+				transition_probability(i,j) = downtick_prob;
+			if (j == i)
+				transition_probability(i,j) = notick_prob;
+			if (j == (i+1))
+				transition_probability(i,j) = uptick_prob;
+		}			
+	
+	Matrix V_t(2*no_of_divisions+1,1);
```

Comment:
see https://google.github.io/styleguide/cppguide.html#Naming for name convention